### PR TITLE
Feature/edit events

### PR DESCRIPTION
**Pull request by natyork**
https://www.pivotaltracker.com/story/show/140278149

This PR updates events controller and associated views to allow an admin to edit/update, or delete an event.

Upon event deletion, any associated presentations are also deleted. A conditional confirmation message for event deletion warns user that presentations for this event will also be deleted.

For views/actions currently limited to admin, a 500 error is displayed if the user is not a logged in admin.

### CODE REVIEW

File: app/views/events/edit.html.erb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+<h1>Edit Event</h1>
+
+<%= render 'form' %>
+
+<%= link_to 'Back', events_path %>
```
Comment (jacquesporveau) [inline app/views/events/edit.html.erb]: Two things:

1.) This commit message doesn't give any hints that we are creating a view. I don't think it should be split into another commit because we would have tests broken on this commit but maybe we could mention it in our PR message.

2.) Linking to back. We don't know how the user got here we shouldn't assume they came from `/events/:id`

File: spec/controllers/events_controller_spec.rb [hunk 2]
```diff
@@ -68,6 +98,16 @@
       end
     end
 
+    describe 'PATCH #update' do
+      let(:event) { FactoryGirl.create(:event) }
+      subject { patch :update, params: { id: event.id, event: attr } }
+      let(:attr) { { name: 'New Name' } }
```
Comment (jacquesporveau) [inline spec/controllers/events_controller_spec.rb]: Super minor but I think it's nice to keep the let statements grouped together. I've kind of been adopting the practice of having the subject above all the let statements just because it's more important than regular variables and we always know where to find it.

File: spec/controllers/events_controller_spec.rb [hunk 3]
```diff
@@ -76,6 +76,17 @@
       end
     end
 
+    describe 'DELETE #destroy' do
+      let!(:event) { FactoryGirl.create(:event) }
```
Comment (jacquesporveau) [inline spec/controllers/events_controller_spec.rb]: Do we need a bang after this let?

File: app/models/event.rb [hunk 4]
```diff
@@ -2,7 +2,7 @@
 class Event < ApplicationRecord
   NOTIFICATION_PERIOD = 1.day
   MAXIMUM_DURATION = Figaro.env.maximum_event_duration.to_i
-  has_many :presentations
+  has_many :presentations, dependent: :destroy
```
Comment (jacquesporveau) [inline app/models/event.rb:5]: Would like to see a spec showing this works.

File: app/views/events/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<h1>Events</h1>
+
+<%= link_to 'New Event', new_event_path %>
+
+<table>
+  <thead>
+    <tr>
+      <th>Event</th>
+    </tr>
```
Comment (jacquesporveau) [inline app/views/events/index.html.erb]: The event partial that you are loading has two `<td>` tags. I would expect there to be two `<th>` tags.

File: app/views/events/show.html.erb [hunk 6]
```diff
@@ -9,3 +9,6 @@
   <dt>Location</dt>
   <dd><%= @event.location %></dd>
 </dl>
+
+<%= link_to 'Edit', edit_event_path(@event) %>
+<%= render partial: 'delete-link', locals: { event: @event } %>
```
Comment (jacquesporveau) [inline app/views/events/show.html.erb:16]: This commit is out of order. The delete link partial hasn't been created yet.

**Review by jacquesporveau — CHANGES_REQUESTED** [review]

File: app/views/events/show.html.erb (hunk 6 above)
Comment (natyork) [reply app/views/events/show.html.erb:16]: The delete link partial was created in the previous commit (Add Delete links to #index).

File: spec/controllers/events_controller_spec.rb [hunk 7]
```diff
@@ -6,6 +6,14 @@
 
     before { mock_pundit_user_as(admin_person) }
 
+    describe 'GET #edit' do
+      let(:event) { FactoryGirl.create(:event) }
+      subject { get :edit, params: { id: event } }
```
Comment (harmonymjb) [inline spec/controllers/events_controller_spec.rb]: Should this be `event.id`?

File: spec/controllers/events_controller_spec.rb [hunk 8]
```diff
@@ -68,6 +76,15 @@
       end
     end
 
+    describe 'GET #edit' do
+      let(:event) { FactoryGirl.create(:event) }
+      subject { get :edit, params: { id: event } }
```
Comment (harmonymjb) [inline spec/controllers/events_controller_spec.rb]: Should this be `event.id`?

File: app/controllers/events_controller.rb [hunk 9]
```diff
@@ -30,6 +30,18 @@ def create
...
 
+  def update
+    @event = Event.find(params[:id])
+    authorize @event
+
+    if @event.update(event_params)
+      flash[:notice] = 'Event was successfully updated.'
+      redirect_to @event
+    else
+      render 'edit'
```
Comment (harmonymjb) [inline app/controllers/events_controller.rb:45]: It might be a good idea to notify the user somehow if the update fails.

File: app/views/events/show.html.erb (hunk 6 above)
Comment (jacquesporveau) [reply app/views/events/show.html.erb:16]: Oh weird it was out of order earler :confused:

**Review by harmonymjb — CHANGES_REQUESTED** [review]
Changes look good, just a couple of questions.

Also, I'm seeing `Unrecognized author. If this is you, make sure natalie@freerunningtechnologies.com is associated with your account. Click to add email addresses in your account settings.` next to your name on each commit. You might need to double check your git settings.

**Review by fredericboivin — APPROVED** [review]
Looks good to me outside of what Graham mentioned already 👍

File: spec/controllers/events_controller_spec.rb (hunk 7 above)
Comment (natyork) [reply spec/controllers/events_controller_spec.rb]: Yes :) Even though this works, I like that using event.id is more explicit.

File: app/controllers/events_controller.rb (hunk 9 above)
Comment (natyork) [reply app/controllers/events_controller.rb:45]: Totally agree and currently if there are errors in updating/creating events they get rendered via the `app/views/events/_errors.html.erb` from the `app/views/events/_form.html.erb` partial (which in turn is rendered from new & edit views.  That pattern already existed in the project so I carried it forward for the addition of edit/update actions. If you can recommend a better approach for this, I think it would be a good idea to create a new story to update this throughout the project.

Comment (natyork) [reply app/controllers/events_controller.rb:45]: Also I realized that the flash notice on success wasn't actually being rendered anywhere so I added that into events#show view.

**Comment by natyork** [comment]
Updates per @omnistegan comments.

**Review by harmonymjb — APPROVED** [review]

**Review by jacquesporveau — APPROVED** [review]
